### PR TITLE
Add support for organizations and organizational units for EC2 image builder distribution configuration

### LINKED ISSUE #21929: `aws_imagebuilder_distribution_configuration`: Add support for sharing AMIs with Organizations and OUs
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or other comments that do not add relevant new information or questions, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Description

Amazon EC2 Image Builder [has added the ability](https://aws.amazon.com/about-aws/whats-new/2021/11/ec2-image-builder-sharing-amis-aws-organization-units/) to share Amazon Machine Images (AMIs) with AWS Organizations and Organizational Units (OUs) (rather than just individual accounts) in the image distribution phase of the build process.  This means we no longer have to update AMI permissions for individual AWS accounts every time a new account is added.

Extend the existing `aws_imagebuilder_distribution_configuration` [resource](https://registry.terraform.io/providers/hashicorp/aws/latest/docs/resources/imagebuilder_infrastructure_configuration) and [data source](https://registry.terraform.io/providers/hashicorp/aws/latest/docs/data-sources/imagebuilder_distribution_configuration) to enable the use of this capability within Terraform.

### New or Affected Resource(s)

- [`aws_imagebuilder_distribution_configuration`](https://registry.terraform.io/providers/hashicorp/aws/latest/docs/resources/imagebuilder_infrastructure_configuration) (affected resource): requires new `organizational_unit_arns` and `organization_arns` sub-arguments within the `distribution.ami_distribution_configuration.launch_permission` argument.
- [`aws_imagebuilder_distribution_configuration`](https://registry.terraform.io/providers/hashicorp/aws/latest/docs/data-sources/imagebuilder_distribution_configuration) (affected data source): requires new `organizational_unit_arns` and `organization_arns` sub-attributes within the `distribution.ami_distribution_configuration.launch_permission` attribute.

### Potential Terraform Configuration

```hcl
resource "aws_imagebuilder_distribution_configuration" "example" {
  name = "example"

  distribution {
    ami_distribution_configuration {
      ami_tags = {
        CostCenter = "IT"
      }

      name = "example-{{ imagebuilder:buildDate }}"

      launch_permission {
        organizational_unit_arns = ["arn:aws:organizations::123456789123:ou/o-foobar123/ou-1foo-2foobar"],
        organization_arns = ["arn:aws:organizations::123456789123:organization/o-foobar123"]
      }
    }

    region = "us-east-1"
  }
}
```

### Requires

- `aws-sdk-go` version 1.42.12: #21939

### References

- AWS Announcement: https://aws.amazon.com/about-aws/whats-new/2021/11/ec2-image-builder-sharing-amis-aws-organization-units/
- User Guide: https://docs.aws.amazon.com/transfer/latest/userguide/transfer-workflows.html
- Relevant EC2 Image Builder API Docs:
  -  `CreateDistributionConfiguration`: https://docs.aws.amazon.com/imagebuilder/latest/APIReference/API_CreateDistributionConfiguration.html
  -  `LaunchPermissionConfiguration`:
https://docs.aws.amazon.com/imagebuilder/latest/APIReference/API_LaunchPermissionConfiguration.html
- Related issue for adding similar support for AMIs via `aws_ami_launch_permission`: #21617

### PR DESCRIPTION
**Pull request by lvthillo**
### Community Note

* Please vote on this pull request by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original pull request comment to help the community and maintainers prioritize this request
* Please do not leave "+1" or other comments that do not add relevant new information or questions, they generate extra noise for pull request followers and do not help prioritize the request

Closes #21929

Output from acceptance testing:

```
$ make testacc TESTS=TestAccImageBuilderDistributionConfiguration_DistributionAMIDistribution PKG=imagebuilder
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./internal/service/imagebuilder/... -v -count 1 -parallel 20 -run='TestAccImageBuilderDistributionConfiguration_DistributionAMIDistribution' -timeout 180m
=== RUN   TestAccImageBuilderDistributionConfiguration_DistributionAMIDistribution_amiTags
=== PAUSE TestAccImageBuilderDistributionConfiguration_DistributionAMIDistribution_amiTags
=== RUN   TestAccImageBuilderDistributionConfiguration_DistributionAMIDistribution_description
=== PAUSE TestAccImageBuilderDistributionConfiguration_DistributionAMIDistribution_description
=== RUN   TestAccImageBuilderDistributionConfiguration_DistributionAMIDistribution_kmsKeyID
=== PAUSE TestAccImageBuilderDistributionConfiguration_DistributionAMIDistribution_kmsKeyID
=== RUN   TestAccImageBuilderDistributionConfiguration_DistributionAMIDistributionLaunchPermission_userGroups
=== PAUSE TestAccImageBuilderDistributionConfiguration_DistributionAMIDistributionLaunchPermission_userGroups
=== RUN   TestAccImageBuilderDistributionConfiguration_DistributionAMIDistributionLaunchPermission_userIDs
=== PAUSE TestAccImageBuilderDistributionConfiguration_DistributionAMIDistributionLaunchPermission_userIDs
=== RUN   TestAccImageBuilderDistributionConfiguration_DistributionAMIDistributionLaunchPermission_organizationArns
--- PASS: TestAccImageBuilderDistributionConfiguration_DistributionAMIDistributionLaunchPermission_organizationArns (28.94s)
=== RUN   TestAccImageBuilderDistributionConfiguration_DistributionAMIDistributionLaunchPermission_organizationalUnitArns
--- PASS: TestAccImageBuilderDistributionConfiguration_DistributionAMIDistributionLaunchPermission_organizationalUnitArns (32.56s)
=== RUN   TestAccImageBuilderDistributionConfiguration_DistributionAMIDistribution_name
=== PAUSE TestAccImageBuilderDistributionConfiguration_DistributionAMIDistribution_name
=== RUN   TestAccImageBuilderDistributionConfiguration_DistributionAMIDistribution_targetAccountIDs
=== PAUSE TestAccImageBuilderDistributionConfiguration_DistributionAMIDistribution_targetAccountIDs
=== CONT  TestAccImageBuilderDistributionConfiguration_DistributionAMIDistribution_amiTags
=== CONT  TestAccImageBuilderDistributionConfiguration_DistributionAMIDistribution_name
=== CONT  TestAccImageBuilderDistributionConfiguration_DistributionAMIDistributionLaunchPermission_userGroups
=== CONT  TestAccImageBuilderDistributionConfiguration_DistributionAMIDistribution_kmsKeyID
=== CONT  TestAccImageBuilderDistributionConfiguration_DistributionAMIDistribution_targetAccountIDs
=== CONT  TestAccImageBuilderDistributionConfiguration_DistributionAMIDistributionLaunchPermission_userIDs
=== CONT  TestAccImageBuilderDistributionConfiguration_DistributionAMIDistribution_description
--- PASS: TestAccImageBuilderDistributionConfiguration_DistributionAMIDistributionLaunchPermission_userGroups (27.73s)
--- PASS: TestAccImageBuilderDistributionConfiguration_DistributionAMIDistributionLaunchPermission_userIDs (45.55s)
--- PASS: TestAccImageBuilderDistributionConfiguration_DistributionAMIDistribution_name (45.58s)
--- PASS: TestAccImageBuilderDistributionConfiguration_DistributionAMIDistribution_amiTags (45.64s)
--- PASS: TestAccImageBuilderDistributionConfiguration_DistributionAMIDistribution_targetAccountIDs (45.72s)
--- PASS: TestAccImageBuilderDistributionConfiguration_DistributionAMIDistribution_description (45.87s)
--- PASS: TestAccImageBuilderDistributionConfiguration_DistributionAMIDistribution_kmsKeyID (46.70s)
PASS
ok      github.com/hashicorp/terraform-provider-aws/internal/service/imagebuilder       110.438s

...
```

Example usage:
```
resource "aws_imagebuilder_distribution_configuration" "example" {
  name = "example"

  distribution {
    ami_distribution_configuration {
      ami_tags = {
        CostCenter = "IT"
      }

      name = "example-{{ imagebuilder:buildDate }}"

      launch_permission {
        organization_arns = ["arn:aws:organizations::111111111111:organization/o-aze123rty456"]
      }
    }

    region = "eu-west-1"
  }
}
```

### CODE REVIEW

**Comment by sethbacon** [comment]
Would love to see this make a release soon!

**Comment by tolidano** [comment]
Keeping this alive - would like to see this merged ASAP - support for organizations is critical for our use case.

**Comment by MaxCCC** [comment]
@justinretzolk Any chance this will be reviewed soon?

**Comment by sabrinaz-orion** [comment]
This would be really good to have!

**Comment by bassjompi** [comment]
+1 organization support

File: internal/service/imagebuilder/distribution_configuration_test.go [hunk 1]
```diff
@@ -704,6 +767,53 @@ resource "aws_imagebuilder_distribution_configuration" "test" {
...
+	distribution {
+	  ami_distribution_configuration {
+		launch_permission {
+		  organizational_unit_arns = [aws_organizations_organizational_unit.test.arn]
+		}
+	  }
+	  region = data.aws_region.current.name
+	}
+  }
+  `, rName)
```
Comment (kamilturek) [inline internal/service/imagebuilder/distribution_configuration_test.go:814]: I think we'll have a bunch of `terrafmt` errors here.

**Review by kamilturek — CHANGES_REQUESTED** [review]
I believe we need to update the documentation and provide a changelog. Besides that, looks good! Let me know if you need any help with it.

File: internal/service/imagebuilder/distribution_configuration_test.go (hunk 1 above)
Comment (lvthillo) [reply internal/service/imagebuilder/distribution_configuration_test.go]: @kamilturek I fixed the `terrafmt` issues + added docs + added changelog.

**Comment by lvthillo** [comment]
> I believe we need to update the documentation and provide a changelog. Besides that, looks good! Let me know if you need any help with it.

I added the changelog entry but got some automated comment that I should remove it. Should I?

**Review by ewbankkit — APPROVED** [review]
LGTM :rocket:.

```console
% make testacc TESTS=TestAccImageBuilderDistributionConfigurationDataSource_ PKG=imagebuilder ACCTEST_PARALLELISM=2
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./internal/service/imagebuilder/... -v -count 1 -parallel 2 -run='TestAccImageBuilderDistributionConfigurationDataSource_'  -timeout 180m
=== RUN   TestAccImageBuilderDistributionConfigurationDataSource_arn
=== PAUSE TestAccImageBuilderDistributionConfigurationDataSource_arn
=== CONT  TestAccImageBuilderDistributionConfigurationDataSource_arn
--- PASS: TestAccImageBuilderDistributionConfigurationDataSource_arn (20.25s)
PASS
ok  	github.com/hashicorp/terraform-provider-aws/internal/service/imagebuilder	23.885s
% make testacc TESTS=TestAccImageBuilderDistributionConfiguration_ PKG=imagebuilder ACCTEST_PARALLELISM=2
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./internal/service/imagebuilder/... -v -count 1 -parallel 2 -run='TestAccImageBuilderDistributionConfiguration_'  -timeout 180m
=== RUN   TestAccImageBuilderDistributionConfiguration_basic
=== PAUSE TestAccImageBuilderDistributionConfiguration_basic
=== RUN   TestAccImageBuilderDistributionConfiguration_disappears
=== PAUSE TestAccImageBuilderDistributionConfiguration_disappears
=== RUN   TestAccImageBuilderDistributionConfiguration_description
=== PAUSE TestAccImageBuilderDistributionConfiguration_description
=== RUN   TestAccImageBuilderDistributionConfiguration_distribution
=== PAUSE TestAccImageBuilderDistributionConfiguration_distribution
=== RUN   TestAccImageBuilderDistributionConfiguration_DistributionAMIDistribution_amiTags
=== PAUSE TestAccImageBuilderDistributionConfiguration_DistributionAMIDistribution_amiTags
=== RUN   TestAccImageBuilderDistributionConfiguration_DistributionAMIDistribution_description
=== PAUSE TestAccImageBuilderDistributionConfiguration_DistributionAMIDistribution_description
=== RUN   TestAccImageBuilderDistributionConfiguration_DistributionAMIDistribution_kmsKeyID
=== PAUSE TestAccImageBuilderDistributionConfiguration_DistributionAMIDistribution_kmsKeyID
=== RUN   TestAccImageBuilderDistributionConfiguration_DistributionAMIDistributionLaunchPermission_userGroups
=== PAUSE TestAccImageBuilderDistributionConfiguration_DistributionAMIDistributionLaunchPermission_userGroups
=== RUN   TestAccImageBuilderDistributionConfiguration_DistributionAMIDistributionLaunchPermission_userIDs
=== PAUSE TestAccImageBuilderDistributionConfiguration_DistributionAMIDistributionLaunchPermission_userIDs
=== RUN   TestAccImageBuilderDistributionConfiguration_DistributionAMIDistributionLaunchPermission_organizationArns
--- PASS: TestAccImageBuilderDistributionConfiguration_DistributionAMIDistributionLaunchPermission_organizationArns (21.91s)
=== RUN   TestAccImageBuilderDistributionConfiguration_DistributionAMIDistributionLaunchPermission_organizationalUnitArns
--- PASS: TestAccImageBuilderDistributionConfiguration_DistributionAMIDistributionLaunchPermission_organizationalUnitArns (21.93s)
=== RUN   TestAccImageBuilderDistributionConfiguration_DistributionAMIDistribution_name
=== PAUSE TestAccImageBuilderDistributionConfiguration_DistributionAMIDistribution_name
=== RUN   TestAccImageBuilderDistributionConfiguration_DistributionAMIDistribution_targetAccountIDs
=== PAUSE TestAccImageBuilderDistributionConfiguration_DistributionAMIDistribution_targetAccountIDs
=== RUN   TestAccImageBuilderDistributionConfiguration_DistributionContainerDistribution_containerTags
=== PAUSE TestAccImageBuilderDistributionConfiguration_DistributionContainerDistribution_containerTags
=== RUN   TestAccImageBuilderDistributionConfiguration_DistributionContainerDistribution_description
=== PAUSE TestAccImageBuilderDistributionConfiguration_DistributionContainerDistribution_description
=== RUN   TestAccImageBuilderDistributionConfiguration_DistributionContainerDistribution_targetRepository
=== PAUSE TestAccImageBuilderDistributionConfiguration_DistributionContainerDistribution_targetRepository
=== RUN   TestAccImageBuilderDistributionConfiguration_Distribution_launchTemplateConfiguration
=== PAUSE TestAccImageBuilderDistributionConfiguration_Distribution_launchTemplateConfiguration
=== RUN   TestAccImageBuilderDistributionConfiguration_Distribution_licenseARNs
=== PAUSE TestAccImageBuilderDistributionConfiguration_Distribution_licenseARNs
=== RUN   TestAccImageBuilderDistributionConfiguration_tags
=== PAUSE TestAccImageBuilderDistributionConfiguration_tags
=== CONT  TestAccImageBuilderDistributionConfiguration_basic
=== CONT  TestAccImageBuilderDistributionConfiguration_DistributionAMIDistribution_name
--- PASS: TestAccImageBuilderDistributionConfiguration_basic (17.67s)
=== CONT  TestAccImageBuilderDistributionConfiguration_DistributionContainerDistribution_targetRepository
--- PASS: TestAccImageBuilderDistributionConfiguration_DistributionAMIDistribution_name (29.77s)
=== CONT  TestAccImageBuilderDistributionConfiguration_tags
--- PASS: TestAccImageBuilderDistributionConfiguration_DistributionContainerDistribution_targetRepository (29.04s)
=== CONT  TestAccImageBuilderDistributionConfiguration_Distribution_licenseARNs
    acctest.go:788: skipping tests; missing IAM service-linked role /aws-service-role/license-manager.amazonaws.com. Please create the role and retry
--- SKIP: TestAccImageBuilderDistributionConfiguration_Distribution_licenseARNs (0.20s)
=== CONT  TestAccImageBuilderDistributionConfiguration_Distribution_launchTemplateConfiguration
--- PASS: TestAccImageBuilderDistributionConfiguration_tags (40.87s)
=== CONT  TestAccImageBuilderDistributionConfiguration_DistributionAMIDistribution_description
--- PASS: TestAccImageBuilderDistributionConfiguration_Distribution_launchTemplateConfiguration (31.10s)
=== CONT  TestAccImageBuilderDistributionConfiguration_DistributionAMIDistributionLaunchPermission_userIDs
--- PASS: TestAccImageBuilderDistributionConfiguration_DistributionAMIDistribution_description (29.16s)
=== CONT  TestAccImageBuilderDistributionConfiguration_DistributionAMIDistributionLaunchPermission_userGroups
--- PASS: TestAccImageBuilderDistributionConfiguration_DistributionAMIDistributionLaunchPermission_userIDs (28.99s)
=== CONT  TestAccImageBuilderDistributionConfiguration_DistributionAMIDistribution_kmsKeyID
--- PASS: TestAccImageBuilderDistributionConfiguration_DistributionAMIDistributionLaunchPermission_userGroups (17.73s)
=== CONT  TestAccImageBuilderDistributionConfiguration_DistributionContainerDistribution_containerTags
--- PASS: TestAccImageBuilderDistributionConfiguration_DistributionAMIDistribution_kmsKeyID (32.80s)
=== CONT  TestAccImageBuilderDistributionConfiguration_DistributionAMIDistribution_targetAccountIDs
--- PASS: TestAccImageBuilderDistributionConfiguration_DistributionContainerDistribution_containerTags (29.76s)
=== CONT  TestAccImageBuilderDistributionConfiguration_DistributionContainerDistribution_description
--- PASS: TestAccImageBuilderDistributionConfiguration_DistributionAMIDistribution_targetAccountIDs (29.54s)
=== CONT  TestAccImageBuilderDistributionConfiguration_DistributionAMIDistribution_amiTags
--- PASS: TestAccImageBuilderDistributionConfiguration_DistributionContainerDistribution_description (29.45s)
=== CONT  TestAccImageBuilderDistributionConfiguration_description
--- PASS: TestAccImageBuilderDistributionConfiguration_DistributionAMIDistribution_amiTags (32.50s)
=== CONT  TestAccImageBuilderDistributionConfiguration_distribution
--- PASS: TestAccImageBuilderDistributionConfiguration_description (30.73s)
=== CONT  TestAccImageBuilderDistributionConfiguration_disappears
--- PASS: TestAccImageBuilderDistributionConfiguration_disappears (13.02s)
--- PASS: TestAccImageBuilderDistributionConfiguration_distribution (20.25s)
PASS
ok  	github.com/hashicorp/terraform-provider-aws/internal/service/imagebuilder	269.422s
```

**Comment by ewbankkit** [comment]
@lvthillo Thanks for the contribution :tada: :clap:.

**Comment by tiendungitd** [comment]
I found issue https://github.com/hashicorp/terraform-provider-aws/issues/21929 released share AMI to organization, but have to use **awscc** provider
Does it the same with this pull request? I just want to ensure that this function also available in AWS provider.

**Comment by lvthillo** [comment]
@tiendungitd Yes it does the same. It's now released in provider version v4.7.0
- https://github.com/hashicorp/terraform-provider-aws/releases/tag/v4.7.0
- https://registry.terraform.io/providers/hashicorp/aws/latest/docs/resources/imagebuilder_distribution_configuration#launch_permission

**Pull request merged**